WalletWasabi/Blockchain/Analysis/Clustering/SmartLabel.cs

```diff
@@ -1,153 +1,168 @@
+	public bool Equals(string? other) =>
+		ToString().Equals(other);
 
-	public bool Equals(SmartLabel? other, IEqualityComparer<string> comparer) => AreEquivalent(this, other, comparer);
+	public bool Equals(string? other, StringComparison comparison) =>
+		ToString().Equals(other, comparison);
 
-	public override int GetHashCode() => ((IStructuralEquatable)Labels).GetHashCode(EqualityComparer<string>.Default);
+	public int CompareTo(SmartLabel other) =>
+		Labels.SequenceCompareTo(other.Labels);
 
-	private static bool AreEquivalent(SmartLabel? x, SmartLabel? y, IEqualityComparer<string>? comparer = null)
+	public int CompareTo(SmartLabel other, IComparer<string> comparer)
 	{
-		if (x is null)
+		if (comparer is null)
 		{
-			if (y is null)
-			{
-				return true;
-			}
-			else
-			{
-				return false;
-			}
+			return CompareTo(other);
 		}
-		else
+
+		// The following code repears what .NET could
+		// if it has SequenceCompareTo accepting a comparer.
+		var thisLabels = Labels;
+		var otherLabels = other.Labels;
+		
+		return CompareToSlow(
+			ref MemoryMarshal.GetReference(thisLabels),
+			thisLabels.Length,
+			ref MemoryMarshal.GetReference(otherLabels),
+			otherLabels.Length,
+			comparer);
+
+		static int CompareToSlow(ref string first, int firstLength, ref string second, int secondLength, IComparer<string> comparer)
 		{
-			if (y is null)
+			int minLength = firstLength;
+            if (minLength > secondLength)
 			{
-				return false;
-			}
-			else
-			{
-				return x.Labels.SequenceEqual(y.Labels, comparer);
+                minLength = secondLength;
 			}
+			
+            for (int i = 0; i < minLength; i++)
+            {
+                int result = comparer.Compare(
+					Unsafe.Add(ref first, i),
+					Unsafe.Add(ref second, i));
+					
+                if (result != 0)
+                    return result;
+            }
+			
+            return firstLength.CompareTo(secondLength);
 		}
 	}
 
-	private static bool AreEquivalent(string? x, SmartLabel? y, StringComparison? comparison = null)
-	{
-		if (x is null)
-		{
-			if (y is null)
-			{
-				return true;
-			}
-			else
-			{
-				return false;
-			}
-		}
-		else
-		{
-			if (y is null)
-			{
-				return false;
-			}
-			else
-			{
-				if (comparison is { })
-				{
-					return string.Equals(x, y.LabelString, comparison.Value);
-				}
+	public int CompareTo(string? other) =>
+		string.Compare(ToString(), other);
 
-				return x == y.LabelString;
-			}
-		}
-	}
+	public int CompareTo(string? other, StringComparison comparison) =>
+		string.Compare(ToString(), other, comparison);
 
-	public static bool operator ==(SmartLabel? x, SmartLabel? y) => AreEquivalent(x, y);
+	public ReadOnlySpan<string>.Enumerator GetEnumerator() =>
+		Labels.GetEnumerator();
 
-	public static bool operator ==(string? x, SmartLabel? y) => AreEquivalent(x, y);
+	IEnumerator<string> IEnumerable<string>.GetEnumerator() => GetEnumeratorAllocating();
```

Comment:
I don't see any unit tests for that part.

Reply:
That part wasn't

Reply:
That part wasn't changed, so added no tests on it, but you're right.

Reply:
@wieslawsoltes, I fixed it, so the pull request is waiting for another review.